### ISSUE TITLE
Add tests for controlled_access_terms

### ISSUE THREAD

**Issue by dannylamb** [Islandora-CLAW/CLAW#901]
@seth-shaw-unlv has generously contributed https://github.com/Islandora-CLAW/controlled_access_terms to the Foundation.  It's mostly config, but there's three new field types that have been made that need tests.  In theory, one functional test should cover a field, its formatter, and widget.  So we'll need three functional tests to cover the use case of configuring and creating content with these fields.

**Comment by Natkeeran** [comment]
@dannylamb I'll try this one.  Can you please point me to example functional test in other modules that I can follow.

**Comment by dannylamb** [comment]
@Natkeeran Awesome, thank you.  The core islandora module has lots of [functional tests](https://github.com/Islandora-CLAW/islandora/tree/8.x-1.x/tests/src/Functional) to look at.  The [abstract base class](https://github.com/Islandora-CLAW/islandora/blob/8.x-1.x/tests/src/Functional/IslandoraFunctionalTestBase.php) has lots of useful functions to check out for the basic idea of how to work with functional tests.

TL;DR you have to test viewing/editing an object when it's got these complex fields, and assert that certain bits are in the html.  You're probably safe, because if you force the tests to enable the `controlled_access_terms` module, you'll have all that config and don't have to go through a bunch of setup.  Just make the nodes and view their pages.

If it turns out to go deeper than that, let me know and we can figure out what's going on.

**Comment by dannylamb** [comment]
@Natkeeran Check out https://github.com/Islandora-CLAW/islandora/blob/8.x-1.x/tests/src/Functional/LinkHeaderTest.php#L59 for a good example of setting up a node and media with references.

**Comment by Natkeeran** [comment]
@dannylamb 

I am having trouble even setting this up.
```
  public static $modules = ['controlled_access_terms'];
```
Is throwing the following error:

```
1) Drupal\Tests\controlled_access_terms\Functional\CreateFamilyTest::testTheSiteStillWorks
Drupal\Core\Config\UnmetDependenciesException: Configuration objects provided by <em class="placeholder">controlled_access_terms</em> have unmet dependencies: <em class="placeholder">core.entity_form_display.node.cat_geographic_location.default (path), node.type.cat_geographic_location (menu_ui), rdf.mapping.node.cat_geographic_location (rdf)</em>
```

Tried added path, menu_ui, and rdf to modules array, and that did not help.  If you have 15 - 30 mins tomorrow to go over this ticket with me tomorrow, that would be helpful.

**Comment by dannylamb** [comment]
@Natkeeran Sure, we can sort it out.  There's always weirdness setting up these tests for the first time.  Can you push what you have to a fork?

**Comment by Natkeeran** [comment]
@dannylamb @seth-shaw-unlv 

I've pushed some basic tests here: https://github.com/Natkeeran/controlled_access_terms/blob/tests/tests/src/Functional/FieldConfigurationTest.php

The test setups a content type with typed_relation field, then creates a term, and a node of that content type to assert that it can create a node with typed_relation field.

Question: Does this cover field, formatter and widget?

If this approach is ok, I can do something similar for Authority Link.  Danny noted that it creates three field types, but it only creates two: https://github.com/Islandora-CLAW/controlled_access_terms/tree/8.x-1.x/src/Plugin/Field/FieldType.

**Comment by whikloj** [comment]
I was looking at some basic tests of the `EDTFUtils::validate()` function and I have come across some issues.

But I'm not super familiar with this module, so I just wanted to document them here to start before I wreck stuff that I don't understand.

1. [This line](https://github.com/Islandora-CLAW/controlled_access_terms/blob/8.x-1.x/src/EDTFUtils.php#L183) is not valid, but works in Drupal. In a Kernel test (or straight PHP) it throws an error. 🤷‍♂ 
1. If [this preg_match](https://github.com/Islandora-CLAW/controlled_access_terms/blob/8.x-1.x/src/EDTFUtils.php#L185) fails the `parsed_date` is an empty array. Again I'm not clear why it works in the Drupal form, but in a Kernel Test it throws errors because we reference that array continuously after it.
1. These [`array_key_exists` checks](https://github.com/Islandora-CLAW/controlled_access_terms/blob/8.x-1.x/src/EDTFUtils.php#L217) test if the key (`self::MONTH` in this case) doesn't exist and isn't empty. I'm guessing this is just a typo and should be `if (array_key_exists(self::MONTH, $parsed_date) && !empty($parsed_date[self::MONTH])) {`
1. [This if statement](https://github.com/Islandora-CLAW/controlled_access_terms/blob/8.x-1.x/src/EDTFUtils.php#L219-L220) uses the MONTHS_MAP, in which 31 is mapped to `Summer`. So if someone enters the date of `1900-31-01` it is considered valid.

I'm still learning how to use this, so I could be wrong in how this should work.

**Comment by seth-shaw-unlv** [comment]
Thanks, @whikloj. I'll come back to this later, but a few points off the top of my head:

1. There used to be more lines like that (it is a common idiom in Python) but I've removed most of them. It won't be hard to fix.
1. Should probably return a failed message in an empty parsed_date condition.
1. You are probably right.
1. So, yes, '31' is summer is the southern hemisphere according to EDTF. The problem with the example value of 1900-31-01, is that dates with 'Divisions of a year' other than months may not have a day value. So, we need to add a check if days were used with a 'month' value greater than 12.

**Comment by dannylamb** [comment]
Resolved via https://github.com/Islandora-CLAW/controlled_access_terms/commit/fafca12b67d5c9edaaa1ff889906bc910a7c4c7f
